### ISSUE TITLE
Compatibility issue on nodejs upgrading to v0.11.14

### ISSUE THREAD

**Issue by ebraminio** [libuv/libuv#124]
I compiled and bisected a nodejs bug I faced with [node-fastcgi](https://github.com/fbbdev/node-fastcgi/issues/2) and filed this [nodejs bug](https://github.com/joyent/node/issues/8988). As the issue is happening on a [libuv upgrade](https://github.com/joyent/node/commit/a35a212f258812bd178f7f002660a0ddc0bf10b2) on nodejs and as @cjihrig [suggested](https://github.com/joyent/node/issues/8988#issuecomment-69128085) I am bringing the bug here. I double rechecked that the issue is just when I compile the revision after the change but I hope I didn't miss anything. As I said on [previous bug](https://github.com/joyent/node/issues/8988#issuecomment-69116738) also it seems there is some issue on handling unix sockets after this change and I hope at least for the sake of compatibility this will be fixed. I suspect to this two changes ([1](https://github.com/libuv/libuv/commit/9d44d786ada6cf94e1bdcee7d777c790ca712a78) @saghul, [2](https://github.com/libuv/libuv/commit/359d6678930fbb25e612d290a2ff5114e10eda6f) @bnoordhuis) on libuv

**Comment by bnoordhuis** [comment]
Can you explain what the actual is and maybe post a test case that doesn't depend on third-party modules?  Aside, libuv v0.11.14  is ancient history by now; joyent/node@v0.12 is now at 1.0.1 or 1.0.2 (I think), iojs/io.js@v1.x is at 1.2.0.

**Comment by ebraminio** [comment]
Actual thing that is happening is just is hang on browser that tries to access node app hosted on lighttpd fcgi. I wish you could repro it by descripton I've given [here](https://github.com/joyent/node/issues/8988), it just need a 0.10 node, a recent 0.11 node and lighttpd. I can not provide a more simiplified sample as my libuv, fcgi and unix socket experience is low but I thought it would be easier for you guys to find the issue.

**Comment by bnoordhuis** [comment]
I'm rather hesitant to start investigating a test case that has thousands of lines of code of which I know nothing and that are not under my control (namely node-fcgi and lighttpd.)  Can you try to narrow it down?

**Comment by ebraminio** [comment]
You are right. I wish I could.

**Comment by ebraminio** [comment]
You are right. I wish I could.

**Comment by saghul** [comment]
Does the bug still show up with latest iojs? It has a more recent libuv.

Also, FWIW, I don't remember any big changes to Unix socket handling...

**Comment by ebraminio** [comment]
Yes, the same issue is available on io.js.

**Comment by saghul** [comment]
@ebraminio sigh :-/ None of the changes you pinted out sound related to me. I guess the only way to figue out what's wrong is to take the first version of Node that shows the problem, and start to bisect libuv there until you find the revision that made things not work.

Unfortunately the setup is too complicated to eyeball the problem :-(

**Comment by piscisaureus** [comment]
Personally I would suspect https://github.com/libuv/libuv/commit/0d435a5 but I don't know either.

@ebraminio can you bisect as to pinpoint the exact libuv commit?

**Comment by ebraminio** [comment]
Excellent! I double checked, the regression is coming from 0d435a5

**Comment by fbbdev** [comment]
The problem is a tricky interaction between libuv and node in ```uv__server_io``` ([```src/unix/stream.c```](https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c)):

```c
void uv__server_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
  uv_stream_t* stream;
  int err;

  stream = container_of(w, uv_stream_t, io_watcher);
  // ...
  while (uv__stream_fd(stream) != -1) { // LOOP
    // ...
    err = uv__accept(uv__stream_fd(stream)); // ACCEPT
    // ...
    stream->accepted_fd = err;
    stream->connection_cb(stream, 0); // NODE CALLBACK
    // ...
  }
}
```

When ```stream->type``` is ```UV_NAMED_PIPE``` and ```uv__accept``` is called again immediately after ```stream->connection_cb```, which is what the loop is doing, socket data callbacks are never called in javascript. The issue can be solved by breaking the loop when the stream is a pipe.

I also inspected node's code and I believe it's correct and this is a libuv issue, however I could be wrong on this.

**Comment by fbbdev** [comment]
In which branch should I put the fix, v1.x or master?

**Comment by saghul** [comment]
> In which branch should I put the fix, v1.x or master?

v1.x. But we'll need a regression test for the fix.

**Comment by bnoordhuis** [comment]
Is the issue that new connections are coming in faster than libuv can accept them?  What I mean is, is the accept loop effectively turning into an infinite loop?

**Comment by fbbdev** [comment]
@bnoordhuis No, it isn't.
I understood what's going on: ```uv__accept``` should be non-blocking but it blocks when called on pipes. The accept4 system call is called correctly in non-blocking mode, nonetheless it blocks.
I did not spot it before because I tought it was the right behaviour.

**Comment by bnoordhuis** [comment]
@fbbdev There is something opening a pipe in blocking mode?  Not something in libuv itself, I hope?

File descriptors that libuv opens itself are always created with the O_NONBLOCK and O_CLOEXEC flags set.  It sounds as if something is calling uv_pipe_open() with a blocking file descriptor, something you can do from node with `net.createServer({ fd: 42 })`.

**Comment by fbbdev** [comment]
@bnoordhuis Yes, my module (node-fastcgi) is calling ```net.createServer(...).listen({fd: 0})```.
The parent process opens the socket, forks and binds it to fd 0. Both official FastCGI tools and modules for Apache and lighttpd open the socket in blocking mode, so there's no workaround for this.

**Comment by bnoordhuis** [comment]
Right.  I don't see how 0d435a5 could have changed that but the fix should be trivial, just force-set the non-blocking flag in `uv_{pipe,tcp,udp}_open()`.  The documentation states that the user is responsible for that but whatever, node users don't have influence on that and it's a cheap operation.

I don't have time at the moment but maybe @saghul is willing to take a look?

**Comment by saghul** [comment]
@bnoordhuis I'll take a look. Actually, Windows does it already, so this inconsistency is sort of a bug.

**Comment by saghul** [comment]
@ebraminio can you test #134 ?

**Comment by ebraminio** [comment]
:+1: I checked #134 twice using latest io.js and it is fixing the issue. Thank you all.

**Comment by ebraminio** [comment]
Is it possible to backport the patch (or temporary reverting 0d435a5) for v1.0? I heard Node 0.12 will be shipped with libuv v1.0. Thank you again.

**Comment by saghul** [comment]
Node should use 1.2.1 (released just now https://github.com/libuv/libuv/releases/tag/v1.2.1) or float the patch. We are not doing more 1.0 releases, as all 1.x releases are ABI compatible and backwards API compatible.

**Comment by saghul** [comment]
#134 landed and it's part of libuv 1.2.1 (https://github.com/libuv/libuv/releases/tag/v1.2.1), closing.
